File: common/mrtjp/projectred/integration/GatePart.java
package mrtjp.projectred.integration;

import java.util.Arrays;

import mrtjp.projectred.ProjectRedIntegration;
import mrtjp.projectred.api.IConnectable;
import mrtjp.projectred.core.BasicUtils;
import mrtjp.projectred.core.BasicWireUtils;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.nbt.NBTTagCompound;
import net.minecraft.util.MovingObjectPosition;
import net.minecraftforge.common.ForgeDirection;
import codechicken.lib.data.MCDataInput;
import codechicken.lib.data.MCDataOutput;
import codechicken.lib.lighting.LazyLightMatrix;
import codechicken.lib.render.CCRenderState;
import codechicken.lib.render.TextureUtils;
import codechicken.lib.vec.BlockCoord;
import codechicken.lib.vec.Cuboid6;
import codechicken.lib.vec.Rotation;
import codechicken.lib.vec.Transformation;
import codechicken.lib.vec.Vector3;
import codechicken.microblock.FaceMicroClass;
import codechicken.multipart.JCuboidPart;
import codechicken.multipart.JNormalOcclusion;
import codechicken.multipart.NormalOcclusionTest;
import codechicken.multipart.PartMap;
import codechicken.multipart.TFacePart;
import codechicken.multipart.TMultiPart;
import codechicken.multipart.TickScheduler;
import codechicken.multipart.TileMultipart;
import cpw.mods.fml.relauncher.Side;
import cpw.mods.fml.relauncher.SideOnly;

@SuppressWarnings({"rawtypes", "unchecked"})
public abstract class GatePart extends JCuboidPart implements JNormalOcclusion, IConnectable, TFacePart
{
    public static Cuboid6[][] oBoxes = new Cuboid6[6][2];
    
    static
    {
        oBoxes[0][0] = new Cuboid6(1/8D, 0, 0, 7/8D, 1/8D, 1);
        oBoxes[0][1] = new Cuboid6(0, 0, 1/8D, 1, 1/8D, 7/8D);
        for(int s = 1; s < 6; s++) {
            Transformation t = Rotation.sideRotations[s].at(Vector3.center);
            oBoxes[s][0] = oBoxes[0][0].copy().apply(t);
            oBoxes[s][1] = oBoxes[0][1].copy().apply(t);
        }
    }
    
    public byte orientation;
    public byte subID;
    public byte shape;
    
    public int connMap;
    public long schedTime;
    
    public abstract GateLogic getLogic();
    
    public int side() {
        return orientation>>2;
    }
    
    public void setSide(int s) {
        orientation = (byte) (orientation&0x3 | s<<2);
    }
    
    public int rotation() {
        return orientation&0x3;
    }
    
    public void setRotation(int r) {
        orientation = (byte) (orientation&0xFC | r);
    }
    
    public int shape() {
        return shape&0xFF;
    }
    
    public void setShape(int s) {
        shape = (byte) s;
    }
    
    public Transformation rotationT() {
        return Rotation.sideOrientation(side(), rotation()).at(Vector3.center);
    }

    public void onPlaced(EntityPlayer player, int side, int meta) {
        subID = (byte) meta;
        setSide(side^1);
        setRotation(Rotation.getSidedRotation(player, side));
    }
    
    @Override
    public void save(NBTTagCompound tag) {
        tag.setByte("orient", orientation);
        tag.setByte("subID", subID);
        tag.setByte("shape", shape);
        tag.setShort("connMap", (short) connMap);
        tag.setLong("schedTime", schedTime);
    }
    
    @Override
    public void load(NBTTagCompound tag) {
        orientation = tag.getByte("orient");
        subID = tag.getByte("subID");
        shape = tag.getByte("shape");
        connMap = tag.getShort("connMap")&0xFFFF;
        schedTime = tag.getLong("schedTime");
    }
    
    @Override
    public void readDesc(MCDataInput packet) {
        orientation = packet.readByte();
        subID = packet.readByte();
        shape = packet.readByte();
    }
    
    public void read(MCDataInput packet) {
        read(packet, packet.readUByte());
    }

    public void read(MCDataInput packet, int switch_key) {
        if(switch_key == 0) {
            orientation = packet.readByte();
            tile().markRender();
        }
        else if(switch_key == 1) {
            shape = packet.readByte();
            tile().markRender();
        }
    }
    
    @Override
    public void writeDesc(MCDataOutput packet) {
        packet.writeByte(orientation);
        packet.writeByte(subID);
        packet.writeByte(shape);
    }
    
    @Override
    public void scheduleTick(int ticks) {
        if(schedTime < 0) {
            schedTime = TickScheduler.getSchedulerTime(world())+ticks;
        }
    }
        
    private void processScheduled() {
        if(schedTime >= 0 && TickScheduler.getSchedulerTime(world()) >= schedTime) {
            schedTime = -1;
            scheduledTick();
        }
    }
    
    public void onChange() {
        processScheduled();
        getLogic().onChange(this);
    }
    
    @Override
    public void update() {
        if(!world().isRemote)
            processScheduled();
        
        getLogic().onTick(this);
    }
    
    @Override
    public void onPartChanged(TMultiPart part) {
        if(!world().isRemote) {
            if(updateInternalConnections())
                onChange();
        }
    }

    @Override
    public void onNeighborChanged() {
        if (!world().isRemote) {
            if(dropIfCantStay())
                return;
            
            updateExternalConnections();
            onChange();
        }
    }
    
    @Override
    public void onAdded() {
        super.onAdded();
        if(!world().isRemote) {
            getLogic().setup(this);
            updateConnections();
            onChange();
        }
    }

    @Override
    public void onRemoved() {
        super.onRemoved();
        if(!world().isRemote) {
            notifyNeighbors(0xF);
        }
    }

    public boolean canStay()
    {
        BlockCoord pos = new BlockCoord(getTile()).offset(side());
        return BasicWireUtils.canPlaceWireOnSide(world(), pos.x, pos.y, pos.z, ForgeDirection.getOrientation(side() ^ 1), false);
    }
    
    public boolean dropIfCantStay()
    {
        if(!canStay())
        {
            drop();
            return true;
        }
        return false;
    }

    public void drop()
    {
        TileMultipart.dropItem(getItem(), world(), Vector3.fromTileEntityCenter(getTile()));
        tile().remPart(this);
    }
        
    public EnumGate getGateType() {
        return EnumGate.VALID_GATES[subID&0xFF];
    }
    
    protected void updateConnections() {
        updateInternalConnections();
        updateExternalConnections();
    }

    /**
     * Recalculates connections to blocks outside this sapce
     * @return true if a new connection was added or one was removed
     */
    protected boolean updateExternalConnections() {
        int newConn = 0;
        for(int r = 0; r < 4; r++)
        {
            if(connectStraight(r))
                newConn|=0x10<<r;
            else if(connectCorner(r))
                newConn|=1<<r;
        }
        
        if(newConn != (connMap & 0xF000FF))
        {
            int diff = connMap^newConn;
            connMap = (connMap&~0xF000FF)|newConn;
            
            //notify corner disconnections
            for(int r = 0; r < 4; r++)
                if((diff & 1<<r) != 0)
                    notifyCornerChange(r);
            
            return true;
        }
        return false;
    }
    
    /**
     * Recalculates connections to other parts within this space
     * @return true if a new connection was added or one was removed
     */
    protected boolean updateInternalConnections() {
        int newConn = 0;
        for(int r = 0; r < 4; r++)
            if(connectInternal(r))
                newConn|=0x100<<r;
        
        if(newConn != (connMap & 0x10F00))
        {
            connMap = (connMap&~0x10F00)|newConn;
            return true;
        }
        return false;
    }
    
    public boolean connectCorner(int r) {
        int absDir = Rotation.rotateSide(side(), r);
        
        BlockCoord pos = new BlockCoord(getTile());
        pos.offset(absDir);
        
        if(!BasicWireUtils.canConnectThroughCorner(world(), pos, absDir^1, side()))
            return false;
        
        pos.offset(side());
        TileMultipart t = BasicUtils.getMultipartTile(world(), pos);
        if (t != null) {
            TMultiPart tp = t.partMap(absDir^1);
            if (tp instanceof IConnectable) {
                IConnectable conn = (IConnectable) tp;
                int r2 = Rotation.rotationTo(absDir^1, side()^1);
                return canConnectTo(conn, r) && conn.canConnectCorner(r2) &&
                        conn.connectCorner(this, r2);
            }
        }
        
        return false;
    }

    public boolean connectStraight(int r) {
        int absDir = Rotation.rotateSide(side(), r);
        
        BlockCoord pos = new BlockCoord(getTile()).offset(absDir);
        TileMultipart t = BasicUtils.getMultipartTile(world(), pos);
        if (t != null) {
            TMultiPart tp = t.partMap(side());
            if (tp instanceof IConnectable) {
                IConnectable conn = (IConnectable) tp;
                return canConnectTo(conn, r) && conn.connectStraight(this, (r+2)%4);
            }
        }
        
        return false;
    }

    public boolean connectInternal(int r) {
        int absDir = Rotation.rotateSide(side(), r);
        
        if(tile().partMap(PartMap.edgeBetween(absDir, side())) != null)
            return false;
        
        TMultiPart tp = tile().partMap(absDir);
        if (tp instanceof IConnectable) {
            IConnectable conn = (IConnectable) tp;
            return canConnectTo(conn, r) && conn.connectInternal(this, Rotation.rotationTo(absDir, side()));
        }
        
        return false;
    }

    public void notifyCornerChange(int r) {
        int absDir = Rotation.rotateSide(side(), r);
        
        BlockCoord pos = new BlockCoord(getTile()).offset(absDir).offset(side());
        world().notifyBlockOfNeighborChange(pos.x, pos.y, pos.z, getTile().getBlockType().blockID);
    }

    public void notifyStraightChange(int r) {
        int absDir = Rotation.rotateSide(side(), r);
        
        BlockCoord pos = new BlockCoord(getTile()).offset(absDir);
        world().notifyBlockOfNeighborChange(pos.x, pos.y, pos.z, getTile().getBlockType().blockID);
    }

    public boolean maskConnects(int r) {
        return (connMap & 0x111 << r) != 0;
    }
    
    @Override
    public Cuboid6 getBounds() {
        return FaceMicroClass.aBounds()[0x10|side()];
    }
    
    @Override
    public Iterable<Cuboid6> getOcclusionBoxes() {
        return Arrays.asList(oBoxes[side()]);
    }
    
    @Override
    public boolean occlusionTest(TMultiPart npart) {
        return NormalOcclusionTest.apply(this, npart);
    }
    
    @Override
    public float getStrength(MovingObjectPosition hit, EntityPlayer player) {
        return hit.sideHit == 1 ? 1.75f : 1.5f;
    }

    public ItemStack getItem() {
        return getGateType().getItemStack();
    }
    
    @Override
    public ItemStack pickItem(MovingObjectPosition hit) {
        return getItem();
    }

    @Override
    public Iterable<ItemStack> getDrops() {
        return Arrays.asList(getItem());
    }

    @Override
    public boolean activate(EntityPlayer player, MovingObjectPosition hit, ItemStack held) {
        if(getLogic().activate(this, player, held))
            return true;
        
        if (held != null && held.getItem() == ProjectRedIntegration.itemScrewdriver) {
            if(!world().isRemote) {
                if (player.isSneaking())
                    configure();
                else
                    rotate();
                held.damageItem(1, player);
            }
            
            return true;
        }
        
        return false;
    }

    public void configure() {
        boolean changed = getLogic().cycleShape(this);
        if(changed) {
            updateConnections();
            tile().markDirty();
            sendShapeUpdate();
            notifyNeighbors(0xF);
            onChange();
        }
    }
    
    public void rotate() {
        setRotation((rotation()+1)%4);
        
        updateConnections();
        tile().markDirty();
        sendOrientationUpdate();
        notifyNeighbors(0xF);
        onChange();
    }
    
    public MCDataOutput getWriteStream(int switch_key) {
        return tile().getWriteStream(this).writeByte(switch_key);
    }
    
    public void sendShapeUpdate() {
        getWriteStream(1).writeByte(shape);
    }
    
    public void sendOrientationUpdate() {
        getWriteStream(0).writeByte(orientation);
    }
    
    /**
     * Notify neighbor blocks
     * @param mask A bitmask of absolute rotation sides to notify
     */
    public void notifyNeighbors(int mask) {
        for(int r = 0; r < 4; r++)
            if((connMap & 1<<r) != 0)
                notifyCornerChange(r);
            else if((connMap & 0x10<<r) != 0)
                notifyStraightChange(r);
    }
    
    public void scheduledTick() {
        getLogic().scheduledTick(this);
    }
    
    @Override
    public boolean connectCorner(IConnectable part, int r)
    {
        if(canConnectTo(part, r))
        {
            connMap|=0x1<<r;
            return true;
        }
        return false;
    }

    @Override
    public boolean connectStraight(IConnectable part, int r)
    {
        if(canConnectTo(part, r))
        {
            connMap|=0x10<<r;
            return true;
        }
        return false;
    }
    
    @Override
    public boolean connectInternal(IConnectable part, int r)
    {
        if(r < 0)
            return false;
        
        if(canConnectTo(part, r))
        {
            connMap|=0x100<<r;
            return true;
        }
        return false;
    }
    
    @Override
    public boolean canConnectCorner(int r) {
        return false;
    }
    
    public int toInternal(int absRot) {
        return (absRot+6-rotation())%4;
    }
    
    public int toAbsolute(int r) {
        return (r+rotation()+2)%4;
    }
    
    public static int shiftMask(int mask, int r) {
        return (mask<<r | mask>>(4-r)) & 0xF;
    }
    
    public static int flipMaskZ(int mask) {
        return mask & 5 | mask<<2 & 8 | mask>>2 & 2;
    }
    
    public int toAbsoluteMask(int mask) {
        return shiftMask(mask, toAbsolute(0));
    }
    
    public int toInternalMask(int mask) {
        return shiftMask(mask, toInternal(0));
    }
    
    public int relRot(int side) {
        return toInternal(Rotation.rotationTo(side(), side));
    }
    
    public boolean canConnectTo(IConnectable part, int r) {
        return getLogic().canConnectTo(this, part, toInternal(r));
    }
    
    @Override
    @SideOnly(Side.CLIENT)
    public void renderStatic(Vector3 pos, LazyLightMatrix olm, int pass) {
        if(pass == 0) {
            TextureUtils.bindAtlas(0);
            CCRenderState.reset();
            CCRenderState.setBrightness(world(), x(), y(), z());
            CCRenderState.useModelColours(true);
            RenderGate.renderStatic(this);
            CCRenderState.setColour(-1);
        }
    }
    
    @Override
    @SideOnly(Side.CLIENT)
    public void renderDynamic(Vector3 pos, float frame, int pass) {
        if(pass == 0)
            TextureUtils.bindAtlas(0);
            RenderGate.renderDynamic(this, pos, frame);
    }
    
    public int getSlotMask() {
        return 1<<side();
    }
    
    @Override
    public int redstoneConductionMap() {
        return 0;
    }
    
    @Override
    public boolean solid(int side) {
        return false;
    }
}

File: src/mrtjp/projectred/fabrication/gatepart.scala
/*
 * Copyright (c) 2015.
 * Created by MrTJP.
 * All rights reserved.
 */
package mrtjp.projectred.fabrication

import codechicken.lib.data.{MCDataInput, MCDataOutput}
import codechicken.lib.vec.Transformation
import cpw.mods.fml.relauncher.{Side, SideOnly}
import mrtjp.core.util.Enum
import mrtjp.projectred.integration.GateDefinition.GateDef
import net.minecraft.nbt.NBTTagCompound

abstract class GateICPart extends CircuitPart with TConnectableICPart with TICOrient with IGuiCircuitPart
{
    private var gateSubID:Byte = 0
    private var gateShape:Byte = 0

    var schedTime = 0L
    var schedDigital = false

    def getLogic[T]:T
    def getLogicPrimitive = getLogic[ICGateLogic[GateICPart]]

    def subID = gateSubID&0xFF

    def shape = gateShape&0xFF
    def setShape(s:Int){ gateShape = s.toByte }

    def preparePlacement(rot:Int, meta:Int)
    {
        gateSubID = meta.toByte
        setRotation(rot)
    }

    override def save(tag:NBTTagCompound)
    {
        tag.setByte("orient", orientation)
        tag.setByte("subID", gateSubID)
        tag.setByte("shape", gateShape)
        tag.setByte("connMap", connMap)
        tag.setLong("schedTime", schedTime)
    }

    override def load(tag:NBTTagCompound)
    {
        orientation = tag.getByte("orient")
        gateSubID = tag.getByte("subID")
        gateShape = tag.getByte("shape")
        connMap = tag.getByte("connMap")
        schedTime = tag.getLong("schedTime")
    }

    override def writeDesc(out:MCDataOutput)
    {
        out.writeByte(orientation)
        out.writeByte(gateSubID)
        out.writeByte(gateShape)
    }

    override def readDesc(in:MCDataInput)
    {
        orientation = in.readByte()
        gateSubID = in.readByte()
        gateShape = in.readByte()
    }

    override def read(in:MCDataInput, key:Int) = key match
    {
        case 1 => orientation = in.readByte()
        case 2 => gateShape = in.readByte()
        case _ => super.read(in, key)
    }

    override def readClientPacket(in:MCDataInput)
    {
        readClientPacket(in, in.readUByte())
    }

    def readClientPacket(in:MCDataInput, key:Int) = key match
    {
        case 0 => rotate()
        case 1 => configure()
        case 2 => getLogicPrimitive.activate(this)
        case _ =>
    }

    override def canConnectPart(part:CircuitPart, r:Int) =
        getLogicPrimitive.canConnectTo(this, part, toInternal(r))

    override def scheduledTick()
    {
        getLogicPrimitive.scheduledTick(this)
    }

    override def scheduleTick(ticks:Int)
    {
        if (ticks == 0) scheduleDigitalTick()
        else if (schedTime < 0) schedTime = world.network.getWorld.getTotalWorldTime+ticks
    }

    def processScheduled()
    {
        if (schedTime >= 0 && world.network.getWorld.getTotalWorldTime >= schedTime)
        {
            schedTime = -1
            scheduledTick()
        }
    }

    def scheduleDigitalTick()
    {
        schedDigital = true
    }

    var iter = 0
    def processScheduledDigital()
    {
        while(schedDigital && iter < 3) //recursion control
        {
            schedDigital = false
            iter += 1
            scheduledTick()
        }
    }

    def onChange()
    {
        processScheduled()
        getLogicPrimitive.onChange(this)
        processScheduledDigital()
    }

    override def update()
    {
        if (!world.network.isRemote)
        {
            processScheduled()
            iter = 0
            processScheduledDigital()
        }
        getLogicPrimitive.onTick(this)
    }

    override def onNeighborChanged()
    {
        if (!world.network.isRemote)
        {
            updateConns()
            onChange()
        }
    }

    override def onAdded()
    {
        super.onAdded()
        if (!world.network.isRemote)
        {
            getLogicPrimitive.setup(this)
            updateConns()
            onChange()
        }
    }

    override def onRemoved()
    {
        super.onRemoved()
        if (!world.network.isRemote) notify(0xF)
    }

    def configure()
    {
        if (getLogicPrimitive.cycleShape(this))
        {
            updateConns()
            world.network.markDirty()
            sendShapeUpdate()
            notify(0xF)
            onChange()
        }
    }

    def rotate()
    {
        setRotation((rotation+1)%4)
        updateConns()
        world.network.markDirty()
        sendOrientUpdate()
        notify(0xF)
        onChange()
    }

    def sendShapeUpdate()
    {
        writeStreamOf(2).writeByte(gateShape)
    }

    def sendOrientUpdate()
    {
        writeStreamOf(1).writeByte(orientation)
    }

    @SideOnly(Side.CLIENT)
    override def renderDynamic(t:Transformation, ortho:Boolean, frame:Float)
    {
        RenderICGate.renderDynamic(this, t, ortho, frame)
    }

    @SideOnly(Side.CLIENT)
    override def getPartName = ICGateDefinition(subID).name

    @SideOnly(Side.CLIENT)
    override def getRolloverData(detailLevel:Int) =
    {
        val s = Seq.newBuilder[String]
        import net.minecraft.util.EnumChatFormatting._
        s ++= getLogicPrimitive.getRolloverData(this, detailLevel)
        super.getRolloverData(detailLevel)++s.result().map(GRAY+_)
    }

    @SideOnly(Side.CLIENT)
    override def createGui = getLogicPrimitive.createGui(this)

    @SideOnly(Side.CLIENT)
    override def onClicked()
    {
        sendClientPacket(_.writeByte(2))
    }

    @SideOnly(Side.CLIENT)
    override def getPickOp =
        CircuitOpDefs.values(CircuitOpDefs.SimpleIO.ordinal+subID).getOp
}

abstract class ICGateLogic[T <: GateICPart]
{
    def canConnectTo(gate:T, part:CircuitPart, r:Int):Boolean

    def cycleShape(gate:T) = false

    def onChange(gate:T)

    def scheduledTick(gate:T)

    def onTick(gate:T){}

    def setup(gate:T){}

    def activate(gate:T){}

    @SideOnly(Side.CLIENT)
    def getRolloverData(gate:T, detailLevel:Int):Seq[String] = Seq.empty

    @SideOnly(Side.CLIENT)
    def createGui(gate:T):CircuitGui = new ICGateGui(gate)
}

trait TComplexGateICPart extends GateICPart
{
    def getLogicComplex = getLogic[TComplexICGateLogic[TComplexGateICPart]]

    def assertLogic()

    abstract override def save(tag:NBTTagCompound)
    {
        super.save(tag)
        getLogicComplex.save(tag)
    }

    abstract override def load(tag:NBTTagCompound)
    {
        super.load(tag)
        assertLogic()
        getLogicComplex.load(tag)
    }

    abstract override def writeDesc(packet:MCDataOutput)
    {
        super.writeDesc(packet)
        getLogicComplex.writeDesc(packet)
    }

    abstract override def readDesc(packet:MCDataInput)
    {
        super.readDesc(packet)
        assertLogic()
        getLogicComplex.readDesc(packet)
    }

    abstract override def read(packet:MCDataInput, key:Int) = key match
    {
        case k if k > 10 =>
            assertLogic() //this may be a net dump part
            getLogicComplex.read(packet, k)
        case _ => super.read(packet, key)
    }

    abstract override def preparePlacement(rot:Int, meta:Int)
    {
        super.preparePlacement(rot, meta)
        assertLogic()
    }
}

trait TComplexICGateLogic[T <: TComplexGateICPart] extends ICGateLogic[T]
{
    def save(tag:NBTTagCompound){}
    def load(tag:NBTTagCompound){}

    def readDesc(packet:MCDataInput){}
    def writeDesc(packet:MCDataOutput){}

    /**
     * Allocated keys > 10
     */
    def read(packet:MCDataInput, key:Int){}
}

object ICGateDefinition extends Enum
{
    type EnumVal = ICGateDef

    import mrtjp.projectred.integration.{GateDefinition => gd}

    val IOSimple = ICGateDef("Simple IO", CircuitPartDefs.IOGate.id)
    val IOAnalog = ICGateDef("Analog IO", CircuitPartDefs.IOGate.id)
    val IOBundled = ICGateDef("Bundled IO", CircuitPartDefs.IOGate.id)

    val OR = ICGateDef("OR gate", CircuitPartDefs.SimpleGate.id, gd.OR)
    val NOR = ICGateDef("NOR gate", CircuitPartDefs.SimpleGate.id, gd.NOR)
    val NOT = ICGateDef("NOT gate", CircuitPartDefs.SimpleGate.id, gd.NOT)
    val AND = ICGateDef("AND gate", CircuitPartDefs.SimpleGate.id, gd.AND)
    val NAND = ICGateDef("NAND gate", CircuitPartDefs.SimpleGate.id, gd.NAND)
    val XOR = ICGateDef("XOR gate", CircuitPartDefs.SimpleGate.id, gd.XOR)
    val XNOR = ICGateDef("XNOR gate", CircuitPartDefs.SimpleGate.id, gd.XNOR)
    val Buffer = ICGateDef("Buffer gate", CircuitPartDefs.SimpleGate.id, gd.Buffer)
    val Multiplexer = ICGateDef("Multiplexer", CircuitPartDefs.SimpleGate.id, gd.Multiplexer)
    val Pulse = ICGateDef("Pulse Former", CircuitPartDefs.SimpleGate.id, gd.Pulse)
    val Repeater = ICGateDef("Repeater", CircuitPartDefs.SimpleGate.id, gd.Repeater)
    val Randomizer = ICGateDef("Randomizer", CircuitPartDefs.SimpleGate.id, gd.Randomizer)
    val SRLatch = ICGateDef("SR Latch", CircuitPartDefs.ComplexGate.id, gd.SRLatch)
    val ToggleLatch = ICGateDef("Toggle Latch", CircuitPartDefs.ComplexGate.id, gd.ToggleLatch)
    val TransparentLatch = ICGateDef("Transparent Latch", CircuitPartDefs.SimpleGate.id, gd.TransparentLatch)
    val Timer = ICGateDef("Timer", CircuitPartDefs.ComplexGate.id, gd.Timer)
    val Sequencer = ICGateDef("Sequencer", CircuitPartDefs.ComplexGate.id, gd.Sequencer)
    val Counter = ICGateDef("Counter", CircuitPartDefs.ComplexGate.id, gd.Counter)
    val StateCell = ICGateDef("State Cell", CircuitPartDefs.ComplexGate.id, gd.StateCell)
    val Synchronizer = ICGateDef("Synchronizer", CircuitPartDefs.ComplexGate.id, gd.Synchronizer)
    val DecRandomizer = ICGateDef("Dec Randomizer", CircuitPartDefs.SimpleGate.id, gd.DecRandomizer)

    case class ICGateDef(unlocal:String, gateType:Int, intDef:GateDef = null) extends Value
    {
        override def name = unlocal
    }
}